### DISCUSSION TITLE
Understanding transformer `X` formats, `X.shape` for 3D numpy format

### DISCUSSION THREAD

**Discussion by haskarb** [alan-turing-institute/sktime#3185] [Q&A]
I'm trying to use sktime transformations to develop my current project.
I'm a bit confused how

`self.n_instances_, _, self.series_length_ = X.shape`

is able to get series length. I picked up this code from [here](https://github.com/alan-turing-institute/sktime/blob/main/sktime/classification/dictionary_based/_boss.py#L181). I have also inherited BaseTransformer.

**Comment by fkiraly** [comment]
Ah, that depends on what the internal format of `X` is - in a transformer, that is specified in the `X_inner_mtype` tag, also see the [transformer extension template](https://github.com/alan-turing-institute/sktime/blob/main/extension_templates/transformer.py).

The strings can be one in a selection, each corresponding to a different data format, see the [datatypes tutorial](https://github.com/alan-turing-institute/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb).

The line you are referring to refers to an `X` in `numpy3D` mtype format, where the three axes correspond to (instance index, variable index, time index), so shape corresponds to (number instances, number variables, series length).

**Reply by haskarb** [reply]
Thanks, Franz.  This is helpful.
